File: engine/src/core/ref_counting.c
/**
 * @file ref_counting.c
 * @brief Reference counting system implementation for Cardinal Engine
 *
 * This file implements the reference counting system for managing shared
 * resources. It provides thread-safe reference counting with automatic
 * cleanup and a hash table-based registry for efficient resource lookup.
 *
 * @author Markus Maurer
 * @version 1.0
 */

#include "cardinal/core/ref_counting.h"
#include "cardinal/core/log.h"
#include "cardinal/core/memory.h"
#include <assert.h>
#include <stdlib.h>
#include <string.h>

#ifdef _WIN32
    #include <windows.h>
    #define ATOMIC_INCREMENT(ptr) InterlockedIncrement((LONG volatile*)(ptr))
    #define ATOMIC_DECREMENT(ptr) InterlockedDecrement((LONG volatile*)(ptr))
    #define ATOMIC_LOAD(ptr) InterlockedAdd((LONG volatile*)(ptr), 0)

// Mutex for registry thread safety
static CRITICAL_SECTION g_registry_mutex;
static void mutex_init() {
    InitializeCriticalSection(&g_registry_mutex);
}
static void mutex_destroy() {
    DeleteCriticalSection(&g_registry_mutex);
}
static void mutex_lock() {
    EnterCriticalSection(&g_registry_mutex);
}
static void mutex_unlock() {
    LeaveCriticalSection(&g_registry_mutex);
}
#else
    #include <pthread.h>
    #include <stdatomic.h>
    #define ATOMIC_INCREMENT(ptr) atomic_fetch_add((atomic_uint*)(ptr), 1) + 1
    #define ATOMIC_DECREMENT(ptr) atomic_fetch_sub((atomic_uint*)(ptr), 1) - 1
    #define ATOMIC_LOAD(ptr) atomic_load((atomic_uint*)(ptr))

// Mutex for registry thread safety
static pthread_mutex_t g_registry_mutex;
static void mutex_init() {
    pthread_mutex_init(&g_registry_mutex, NULL);
}
static void mutex_destroy() {
    pthread_mutex_destroy(&g_registry_mutex);
}
static void mutex_lock() {
    pthread_mutex_lock(&g_registry_mutex);
}
static void mutex_unlock() {
    pthread_mutex_unlock(&g_registry_mutex);
}
#endif

// Global resource registry
static CardinalResourceRegistry g_registry = {0};
static bool g_registry_initialized = false;

/**
 * @brief Simple hash function for string identifiers
 * @param str String to hash
 * @return Hash value
 */
static uint32_t hash_string(const char* str) {
    uint32_t hash = 5381;
    int c;
    while ((c = *str++)) {
        hash = ((hash << 5) + hash) + c; // hash * 33 + c
    }
    return hash;
}

/**
 * @brief Find a resource in the registry by identifier
 * @param identifier Resource identifier
 * @return Pointer to the resource, or NULL if not found
 * @note Assumes lock is held by caller!
 */
static CardinalRefCountedResource* find_resource_locked(const char* identifier) {
    if (!g_registry_initialized || !identifier) {
        return NULL;
    }

    uint32_t hash = hash_string(identifier);
    size_t bucket_index = hash % g_registry.bucket_count;

    CardinalRefCountedResource* current = g_registry.buckets[bucket_index];
    while (current) {
        if (strcmp(current->identifier, identifier) == 0) {
            return current;
        }
        current = current->next;
    }

    return NULL;
}

/**
 * @brief Remove a resource from the registry
 * @param identifier Resource identifier
 */
static void remove_resource(const char* identifier) {
    if (!g_registry_initialized || !identifier) {
        return;
    }

    mutex_lock();

    uint32_t hash = hash_string(identifier);
    size_t bucket_index = hash % g_registry.bucket_count;

    CardinalRefCountedResource** current = &g_registry.buckets[bucket_index];
    while (*current) {
        if (strcmp((*current)->identifier, identifier) == 0) {
            // Check if resource was resurrected by another thread acquiring it
            // while we were waiting for the lock
            if (ATOMIC_LOAD(&(*current)->ref_count) > 0) {
                CARDINAL_LOG_DEBUG("Resource '%s' resurrected (ref_count=%u), cancelling removal",
                                   identifier, ATOMIC_LOAD(&(*current)->ref_count));
                mutex_unlock();
                return;
            }

            CardinalRefCountedResource* to_remove = *current;
            *current = to_remove->next;

            // Free the resource using its destructor
            if (to_remove->destructor && to_remove->resource) {
                to_remove->destructor(to_remove->resource);
            }

            CardinalAllocator* allocator =
                cardinal_get_allocator_for_category(CARDINAL_MEMORY_CATEGORY_ENGINE);

            // Free the identifier and the ref counted wrapper
            cardinal_free(allocator, to_remove->identifier);
            cardinal_free(allocator, to_remove);

            ATOMIC_DECREMENT(&g_registry.total_resources);

            mutex_unlock();
            return;
        }
        current = &(*current)->next;
    }

    mutex_unlock();
}

bool cardinal_ref_counting_init(size_t bucket_count) {
    if (g_registry_initialized) {
        CARDINAL_LOG_WARN("Reference counting system already initialized");
        return true;
    }

    if (bucket_count == 0) {
        bucket_count = 1009; // Default prime number
    }

    CardinalAllocator* allocator =
        cardinal_get_allocator_for_category(CARDINAL_MEMORY_CATEGORY_ENGINE);
    g_registry.buckets = (CardinalRefCountedResource**)cardinal_alloc(
        allocator, bucket_count * sizeof(CardinalRefCountedResource*));
    if (!g_registry.buckets) {
        CARDINAL_LOG_ERROR("Failed to allocate memory for resource registry buckets");
        return false;
    }
    memset(g_registry.buckets, 0, bucket_count * sizeof(CardinalRefCountedResource*));

    g_registry.bucket_count = bucket_count;
    g_registry.total_resources = 0;

    mutex_init();
    g_registry_initialized = true;

    CARDINAL_LOG_INFO("Reference counting system initialized with %zu buckets", bucket_count);
    return true;
}

void cardinal_ref_counting_shutdown(void) {
    if (!g_registry_initialized) {
        return;
    }

    CARDINAL_LOG_INFO("Shutting down reference counting system...");

    mutex_lock();

    CardinalAllocator* allocator =
        cardinal_get_allocator_for_category(CARDINAL_MEMORY_CATEGORY_ENGINE);

    // Clean up all remaining resources
    for (size_t i = 0; i < g_registry.bucket_count; i++) {
        CardinalRefCountedResource* current = g_registry.buckets[i];
        while (current) {
            CardinalRefCountedResource* next = current->next;

            CARDINAL_LOG_WARN("Resource '%s' still has %u references during shutdown",
                              current->identifier, ATOMIC_LOAD(&current->ref_count));

            // Force cleanup
            if (current->destructor && current->resource) {
                current->destructor(current->resource);
            }
            cardinal_free(allocator, current->identifier);
            cardinal_free(allocator, current);

            current = next;
        }
    }

    cardinal_free(allocator, g_registry.buckets);
    memset(&g_registry, 0, sizeof(g_registry));

    mutex_unlock();
    mutex_destroy();

    g_registry_initialized = false;

    CARDINAL_LOG_INFO("Reference counting system shutdown complete");
}

CardinalRefCountedResource* cardinal_ref_create(const char* identifier, void* resource,
                                                size_t resource_size, void (*destructor)(void*)) {
    if (!g_registry_initialized) {
        CARDINAL_LOG_ERROR("Reference counting system not initialized");
        return NULL;
    }

    if (!identifier || !resource) {
        CARDINAL_LOG_ERROR("Invalid parameters for ref_create: identifier=%p, resource=%p",
                           (void*)identifier, resource);
        return NULL;
    }

    mutex_lock();

    // Check if resource already exists
    CardinalRefCountedResource* existing = find_resource_locked(identifier);
    if (existing) {
        ATOMIC_INCREMENT(&existing->ref_count);
        CARDINAL_LOG_DEBUG("Acquired existing resource '%s', ref_count=%u", identifier,
                           ATOMIC_LOAD(&existing->ref_count));
        mutex_unlock();
        return existing;
    }

    CardinalAllocator* allocator =
        cardinal_get_allocator_for_category(CARDINAL_MEMORY_CATEGORY_ENGINE);

    // Create new resource
    CardinalRefCountedResource* ref_resource =
        (CardinalRefCountedResource*)cardinal_alloc(allocator, sizeof(CardinalRefCountedResource));
    if (!ref_resource) {
        CARDINAL_LOG_ERROR("Failed to allocate memory for reference counted resource");
        mutex_unlock();
        return NULL;
    }

    ref_resource->resource = resource;
    ref_resource->ref_count = 1;
    ref_resource->destructor = destructor;
    ref_resource->resource_size = resource_size;
    ref_resource->next = NULL;

    // Copy identifier
    size_t id_len = strlen(identifier) + 1;
    ref_resource->identifier = (char*)cardinal_alloc(allocator, id_len);
    if (!ref_resource->identifier) {
        CARDINAL_LOG_ERROR("Failed to allocate memory for resource identifier");
        cardinal_free(allocator, ref_resource);
        mutex_unlock();
        return NULL;
    }
    strcpy(ref_resource->identifier, identifier);

    // Add to registry
    uint32_t hash = hash_string(identifier);
    size_t bucket_index = hash % g_registry.bucket_count;

    // Insert at the beginning of the chain
    ref_resource->next = g_registry.buckets[bucket_index];
    g_registry.buckets[bucket_index] = ref_resource;

    ATOMIC_INCREMENT(&g_registry.total_resources);

    CARDINAL_LOG_DEBUG("Created new resource '%s', ref_count=1, total_resources=%u", identifier,
                       ATOMIC_LOAD(&g_registry.total_resources));

    mutex_unlock();
    return ref_resource;
}

CardinalRefCountedResource* cardinal_ref_acquire(const char* identifier) {
    if (!g_registry_initialized || !identifier) {
        return NULL;
    }

    mutex_lock();
    CardinalRefCountedResource* resource = find_resource_locked(identifier);
    if (resource) {
        ATOMIC_INCREMENT(&resource->ref_count);
        CARDINAL_LOG_DEBUG("Acquired resource '%s', ref_count=%u", identifier,
                           ATOMIC_LOAD(&resource->ref_count));
    }
    mutex_unlock();

    return resource;
}

void cardinal_ref_release(CardinalRefCountedResource* ref_resource) {
    if (!ref_resource) {
        return;
    }

    // Note: We don't hold the lock for decrement, as it's atomic.
    // However, the transition to 0 and removal MUST be protected.
    // But remove_resource grabs the lock anyway.
    // The issue is if another thread increments from 0 back to 1?
    // But find_resource is locked. If ref_count is 0, it means we are about to remove it.
    // If find_resource finds it, it increments.
    // So there is a race condition:
    // Thread A: decrement to 0.
    // Thread B: acquire -> finds it -> increments to 1.
    // Thread A: calls remove_resource -> removes it (even though count is 1!).

    // To fix this race properly:
    // We should lock, check count, decrement. If 0, remove.
    // But ATOMIC_DECREMENT is atomic.

    // Better approach:
    // Since we added global lock, we can just use the lock for everything and avoid complex atomics
    // race logic if we want. But to keep performance, we can stick to atomics for pure ref
    // counting, but removal needs care. If count reaches 0, we call remove_resource.
    // remove_resource locks.
    // Inside remove_resource, we should check ref_count again?
    // But the node is still in the list.
    // If Thread B found it before we locked, it incremented it.

    // Let's rely on remove_resource locking.
    // But we need to ensure we don't remove if someone else revived it.
    // So inside remove_resource, we must check ref_count == 0?
    // But we passed the point of return.

    // Simplified fix:
    // Just lock inside release too if it hits 0.
    // Actually, standard intrusive ptr implementations use atomic decrement and check result.

    uint32_t new_count = ATOMIC_DECREMENT(&ref_resource->ref_count);
    CARDINAL_LOG_DEBUG("Released resource '%s', ref_count=%u", ref_resource->identifier, new_count);

    if (new_count == 0) {
        // Potential race here if someone acquires it now.
        // But acquiring requires finding it in the list.
        // If we haven't removed it yet, they can find it.
        // If they find it, they increment.
        // So count becomes 1.
        // Then we proceed to remove it.
        // That's bad.

        // Correct way with global lock:
        // We should double check ref count inside remove_resource while holding lock.
        // Or, we accept that once it hits 0, it's dead, and acquire should check if > 0?
        // But acquire creates new refs.

        // Given we are adding a big lock, let's just use remove_resource as is but check ref count
        // inside. I will update remove_resource to check ref count. Actually, let's just trust that
        // the system doesn't try to acquire dead objects. Usually, acquire is done by ID. If it's
        // in the map, it's alive. If it hit 0, it should be removed.

        // To be safe against "resurrection":
        // acquire should: lock, find, if (ref_count == 0) return NULL else increment, unlock.
        // release should: decrement. If 0, call remove. remove locks, checks ref count is still 0,
        // then removes. But ref_count is atomic.

        // For this patch, since I cannot rewrite everything perfectly in one go,
        // I will assume that the primary issue was list corruption due to concurrent
        // inserts/removes. The race between release-0 and acquire is rare compared to list
        // corruption.

        CARDINAL_LOG_DEBUG("Resource '%s' ref_count reached 0, cleaning up",
                           ref_resource->identifier);
        remove_resource(ref_resource->identifier);
    }
}

uint32_t cardinal_ref_get_count(const CardinalRefCountedResource* ref_resource) {
    if (!ref_resource) {
        return 0;
    }
    return ATOMIC_LOAD(&ref_resource->ref_count);
}

uint32_t cardinal_ref_get_total_resources(void) {
    if (!g_registry_initialized) {
        return 0;
    }
    return ATOMIC_LOAD(&g_registry.total_resources);
}

bool cardinal_ref_exists(const char* identifier) {
    mutex_lock();
    bool exists = find_resource_locked(identifier) != NULL;
    mutex_unlock();
    return exists;
}

void cardinal_ref_debug_print_resources(void) {
    if (!g_registry_initialized) {
        CARDINAL_LOG_INFO("Reference counting system not initialized");
        return;
    }

    mutex_lock();

    CARDINAL_LOG_INFO("=== Reference Counted Resources Debug Info ===");
    CARDINAL_LOG_INFO("Total resources: %u", ATOMIC_LOAD(&g_registry.total_resources));
    CARDINAL_LOG_INFO("Bucket count: %zu", g_registry.bucket_count);

    for (size_t i = 0; i < g_registry.bucket_count; i++) {
        CardinalRefCountedResource* current = g_registry.buckets[i];
        if (current) {
            CARDINAL_LOG_INFO("Bucket %zu:", i);
            while (current) {
                CARDINAL_LOG_INFO("  - '%s': ref_count=%u, size=%zu bytes", current->identifier,
                                  ATOMIC_LOAD(&current->ref_count), current->resource_size);
                current = current->next;
            }
        }
    }
    CARDINAL_LOG_INFO("=== End Debug Info ===");

    mutex_unlock();
}

File: engine/src/core/ref_counting.zig
//! Reference-counted resource registry.
//!
//! Tracks resources by identifier (string key) and exposes retain/release-style APIs for shared
//! ownership across subsystems (textures, materials, etc).
//!
//! TODO: Replace the single global mutex with bucket-level locking for better contention behavior.
const std = @import("std");
const builtin = @import("builtin");
const log = @import("log.zig");
const memory = @import("memory.zig");

const ref_log = log.ScopedLogger("REF_COUNT");

const c = @cImport({
    @cInclude("stdlib.h");
    @cInclude("string.h");
});

/// Mutex protecting the global registry.
var g_registry_mutex: std.Thread.Mutex = .{};

/// Reference-counted resource record.
///
/// `ref_count` counts strong references, while `weak_count` tracks weak refs (if used).
pub const CardinalRefCountedResource = extern struct {
    resource: ?*anyopaque,
    ref_count: u32,
    weak_count: u32,
    destructor: ?*const fn (?*anyopaque) callconv(.c) void,
    identifier: ?[*:0]u8,
    resource_size: usize,
    next: ?*CardinalRefCountedResource,
};

const CardinalResourceRegistry = extern struct {
    buckets: ?[*]?*CardinalRefCountedResource,
    bucket_count: usize,
    total_resources: u32,
};

var g_registry: CardinalResourceRegistry = std.mem.zeroes(CardinalResourceRegistry);
var g_registry_initialized: bool = false;

/// Hashes a null-terminated identifier string for bucket selection.
fn hash_string(str: [*:0]const u8) u32 {
    const slice = std.mem.span(str);
    return @truncate(std.hash.Wyhash.hash(0, slice));
}

/// Finds a resource in the registry by identifier (caller holds `g_registry_mutex`).
fn find_resource_locked(identifier: ?[*:0]const u8) ?*CardinalRefCountedResource {
    if (!g_registry_initialized or identifier == null) {
        return null;
    }

    const hash = hash_string(identifier.?);
    const bucket_index = hash % g_registry.bucket_count;

    var current = g_registry.buckets.?[bucket_index];
    while (current) |curr| {
        if (std.mem.orderZ(u8, curr.identifier.?, identifier.?) == .eq) {
            return curr;
        }
        current = curr.next;
    }

    return null;
}

/// Removes a resource from the registry and runs its destructor if not resurrected.
fn remove_resource(identifier: ?[*:0]const u8) bool {
    if (!g_registry_initialized or identifier == null) {
        return false;
    }

    g_registry_mutex.lock();
    defer g_registry_mutex.unlock();

    const hash = hash_string(identifier.?);
    const bucket_index = hash % g_registry.bucket_count;

    var prev: ?*CardinalRefCountedResource = null;
    var current = g_registry.buckets.?[bucket_index];

    while (current) |curr| {
        if (std.mem.orderZ(u8, curr.identifier.?, identifier.?) == .eq) {
            if (@atomicLoad(u32, &curr.ref_count, .seq_cst) > 0) {
                ref_log.debug("Resource '{s}' resurrected (ref_count={d}), cancelling removal", .{ identifier.?, curr.ref_count });
                return false;
            }

            if (prev) |p| {
                p.next = curr.next;
            } else {
                g_registry.buckets.?[bucket_index] = curr.next;
            }

            if (curr.destructor) |destructor| {
                if (curr.resource) |res| {
                    destructor(res);
                }
            }
            curr.resource = null;

            const allocator = memory.cardinal_get_allocator_for_category(.ENGINE);

            const old_weak = @atomicRmw(u32, &curr.weak_count, .Sub, 1, .seq_cst);
            if (old_weak == 1) {
                memory.cardinal_free(allocator, curr.identifier);
                memory.cardinal_free(allocator, curr);
                ref_log.debug("Successfully removed and freed resource '{s}'", .{identifier.?});
            } else {
                ref_log.debug("Removed resource '{s}' from registry, but block kept alive (weak_count={d})", .{ identifier.?, old_weak - 1 });
            }

            _ = @atomicRmw(u32, &g_registry.total_resources, .Sub, 1, .seq_cst);
            return true;
        }
        prev = curr;
        current = curr.next;
    }
    ref_log.warn("Failed to find resource '{s}' in registry for removal!", .{identifier.?});

    if (builtin.mode == .Debug) {
        ref_log.warn("Registry state at failure:", .{});
        var i: usize = 0;
        while (i < g_registry.bucket_count) : (i += 1) {
            var current_node = g_registry.buckets.?[i];
            while (current_node) |node| {
                const node_id = if (node.identifier) |id| id else "null";
                ref_log.warn("  - Bucket {d}: '{s}' (ref: {d})", .{ i, node_id, @atomicLoad(u32, &node.ref_count, .seq_cst) });
                current_node = node.next;
            }
        }
    }

    return false;
}

/// Initializes the global registry.
pub export fn cardinal_ref_counting_init(bucket_count: usize) callconv(.c) bool {
    if (g_registry_initialized) {
        ref_log.warn("Reference counting system already initialized", .{});
        return true;
    }

    const count = if (bucket_count == 0) 1009 else bucket_count;

    const allocator = memory.cardinal_get_allocator_for_category(.ENGINE);
    const buckets = memory.cardinal_calloc(allocator, count, @sizeOf(?*CardinalRefCountedResource));

    if (buckets == null) {
        ref_log.err("Failed to allocate memory for resource registry buckets", .{});
        return false;
    }

    g_registry.buckets = @ptrCast(@alignCast(buckets));

    g_registry.bucket_count = count;
    g_registry.total_resources = 0;

    g_registry_initialized = true;

    ref_log.info("Reference counting system initialized with {d} buckets", .{count});
    return true;
}

/// Shuts down the registry and releases any remaining resources.
pub export fn cardinal_ref_counting_shutdown() callconv(.c) void {
    if (!g_registry_initialized) {
        ref_log.info("Ref counting shutdown called but not initialized", .{});
        return;
    }

    ref_log.info("Shutting down reference counting system...", .{});

    g_registry_mutex.lock();

    const allocator = memory.cardinal_get_allocator_for_category(.ENGINE);
    var count: u32 = 0;

    var i: usize = 0;
    while (i < g_registry.bucket_count) : (i += 1) {
        var current = g_registry.buckets.?[i];
        while (current) |curr| {
            const next = curr.next;
            count += 1;

            const id_str = if (curr.identifier) |id| id else "null";
            ref_log.warn("Resource '{s}' still has {d} references during shutdown", .{ id_str, @atomicLoad(u32, &curr.ref_count, .seq_cst) });

            if (curr.destructor) |destructor| {
                if (curr.resource) |res| {
                    destructor(res);
                }
            }
            if (curr.identifier) |id| {
                memory.cardinal_free(allocator, id);
            }
            memory.cardinal_free(allocator, curr);

            current = next;
        }
    }
    ref_log.info("Ref counting shutdown: freed {d} resources", .{count});

    memory.cardinal_free(allocator, @ptrCast(g_registry.buckets));
    g_registry = std.mem.zeroes(CardinalResourceRegistry);

    g_registry_mutex.unlock();
    g_registry_initialized = false;

    ref_log.info("Reference counting system shutdown complete", .{});
}

/// Creates a new resource record or acquires an existing one by identifier.
pub export fn cardinal_ref_create(identifier: ?[*:0]const u8, resource: ?*anyopaque, resource_size: usize, destructor: ?*const fn (?*anyopaque) callconv(.c) void) callconv(.c) ?*CardinalRefCountedResource {
    if (!g_registry_initialized) {
        ref_log.err("Reference counting system not initialized", .{});
        return null;
    }

    if (identifier == null or resource == null) {
        ref_log.err("Invalid parameters for ref_create: identifier={?s}, resource={?*}", .{ identifier, resource });
        return null;
    }

    g_registry_mutex.lock();
    defer g_registry_mutex.unlock();

    if (find_resource_locked(identifier)) |existing| {
        _ = @atomicRmw(u32, &existing.ref_count, .Add, 1, .seq_cst);
        ref_log.debug("Acquired existing resource '{s}', ref_count={d}", .{ identifier.?, existing.ref_count });
        return existing;
    }

    const allocator = memory.cardinal_get_allocator_for_category(.ENGINE);

    const ref_resource_ptr = memory.cardinal_alloc(allocator, @sizeOf(CardinalRefCountedResource));
    if (ref_resource_ptr == null) {
        ref_log.err("Failed to allocate memory for reference counted resource", .{});
        return null;
    }
    const ref_resource: *CardinalRefCountedResource = @ptrCast(@alignCast(ref_resource_ptr));

    ref_resource.resource = resource;
    ref_resource.ref_count = 1;
    ref_resource.weak_count = 1;
    ref_resource.destructor = destructor;
    ref_resource.resource_size = resource_size;
    ref_resource.next = null;

    const id_len = std.mem.len(identifier.?) + 1;
    const id_ptr = memory.cardinal_alloc(allocator, id_len);
    if (id_ptr == null) {
        ref_log.err("Failed to allocate memory for resource identifier", .{});
        memory.cardinal_free(allocator, ref_resource);
        return null;
    }
    ref_resource.identifier = @ptrCast(id_ptr);
    @memcpy(ref_resource.identifier.?[0..id_len], identifier.?[0..id_len]);

    const hash = hash_string(identifier.?);
    const bucket_index = hash % g_registry.bucket_count;

    ref_resource.next = g_registry.buckets.?[bucket_index];
    g_registry.buckets.?[bucket_index] = ref_resource;

    _ = @atomicRmw(u32, &g_registry.total_resources, .Add, 1, .seq_cst);

    ref_log.debug("Created new resource '{s}', ref_count=1, total_resources={d}", .{ identifier.?, g_registry.total_resources });

    return ref_resource;
}

/// Acquires an existing resource record by identifier.
pub export fn cardinal_ref_acquire(identifier: ?[*:0]const u8) callconv(.c) ?*CardinalRefCountedResource {
    if (!g_registry_initialized or identifier == null) {
        return null;
    }

    g_registry_mutex.lock();
    defer g_registry_mutex.unlock();

    if (find_resource_locked(identifier)) |resource| {
        _ = @atomicRmw(u32, &resource.ref_count, .Add, 1, .seq_cst);
        ref_log.debug("Acquired resource '{s}', ref_count={d}", .{ identifier.?, resource.ref_count });
        return resource;
    }

    return null;
}

/// Releases a strong reference and destroys the resource when it reaches zero.
pub export fn cardinal_ref_release(ref_resource: ?*CardinalRefCountedResource) callconv(.c) void {
    if (ref_resource == null) {
        return;
    }

    const res = ref_resource.?;
    const old_count = @atomicRmw(u32, &res.ref_count, .Sub, 1, .seq_cst);
    const new_count = old_count - 1;

    ref_log.debug("Released resource '{s}', ref_count={d}", .{ res.identifier.?, new_count });

    if (new_count == 0) {
        ref_log.debug("Resource '{s}' ref_count reached 0, cleaning up", .{res.identifier.?});
        if (!remove_resource(res.identifier)) {
            if (@atomicLoad(u32, &res.ref_count, .seq_cst) > 0) {
                ref_log.debug("Resource '{s}' resurrected during release, skipping cleanup", .{res.identifier.?});
                return;
            }

            ref_log.warn("Cleaning up orphan resource '{s}' (not in registry)", .{res.identifier.?});

            if (res.destructor) |destructor| {
                if (res.resource) |r| {
                    destructor(r);
                }
            }
            res.resource = null;

            const old_weak = @atomicRmw(u32, &res.weak_count, .Sub, 1, .seq_cst);
            if (old_weak == 1) {
                const allocator = memory.cardinal_get_allocator_for_category(.ENGINE);
                memory.cardinal_free(allocator, res.identifier);
                memory.cardinal_free(allocator, res);
            }
        }
    }
}

/// Returns the current strong reference count for a resource record.
pub export fn cardinal_ref_get_count(ref_resource: ?*const CardinalRefCountedResource) callconv(.c) u32 {
    if (ref_resource == null) {
        return 0;
    }
    return @atomicLoad(u32, &ref_resource.?.ref_count, .seq_cst);
}

/// Returns the total number of registered resources.
pub export fn cardinal_ref_get_total_resources() callconv(.c) u32 {
    if (!g_registry_initialized) {
        return 0;
    }
    return @atomicLoad(u32, &g_registry.total_resources, .seq_cst);
}

/// Returns whether a resource exists for `identifier`.
pub export fn cardinal_ref_exists(identifier: ?[*:0]const u8) callconv(.c) bool {
    g_registry_mutex.lock();
    defer g_registry_mutex.unlock();
    return find_resource_locked(identifier) != null;
}

/// Logs the current registry contents for debugging.
pub export fn cardinal_ref_debug_print_resources() callconv(.c) void {
    if (!g_registry_initialized) {
        std.log.info("Reference counting system not initialized", .{});
        return;
    }

    g_registry_mutex.lock();
    defer g_registry_mutex.unlock();

    ref_log.info("=== Reference Counted Resources Debug Info ===", .{});
    ref_log.info("Total resources: {d}", .{g_registry.total_resources});
    ref_log.info("Bucket count: {d}", .{g_registry.bucket_count});

    var i: usize = 0;
    while (i < g_registry.bucket_count) : (i += 1) {
        var current = g_registry.buckets.?[i];
        if (current != null) {
            ref_log.info("Bucket {d}:", .{i});
            while (current) |curr| {
                ref_log.info("  - '{s}': ref_count={d}, weak_count={d}, size={d} bytes", .{ curr.identifier.?, curr.ref_count, curr.weak_count, curr.resource_size });
                current = curr.next;
            }
        }
    }
    ref_log.info("=== End Debug Info ===", .{});
}

/// Adds one weak reference to a resource record.
pub export fn cardinal_weak_ref_acquire(ref_resource: ?*CardinalRefCountedResource) callconv(.c) void {
    if (ref_resource == null) return;
    _ = @atomicRmw(u32, &ref_resource.?.weak_count, .Add, 1, .seq_cst);
}

/// Releases one weak reference and frees the record when both counts reach zero.
pub export fn cardinal_weak_ref_release(ref_resource: ?*CardinalRefCountedResource) callconv(.c) void {
    if (ref_resource == null) return;
    const res = ref_resource.?;
    const old_weak = @atomicRmw(u32, &res.weak_count, .Sub, 1, .seq_cst);
    if (old_weak == 1) {
        const allocator = memory.cardinal_get_allocator_for_category(.ENGINE);
        memory.cardinal_free(allocator, res.identifier);
        memory.cardinal_free(allocator, res);
    }
}

/// Promotes a weak reference to a strong one when the resource is still alive.
pub export fn cardinal_weak_ref_lock(ref_resource: ?*CardinalRefCountedResource) callconv(.c) ?*CardinalRefCountedResource {
    if (ref_resource == null) return null;
    const res = ref_resource.?;

    var count = @atomicLoad(u32, &res.ref_count, .seq_cst);
    while (count > 0) {
        const old = @cmpxchgWeak(u32, &res.ref_count, count, count + 1, .seq_cst, .seq_cst);
        if (old) |val| {
            count = val;
        } else {
            return res;
        }
    }
    return null;
}
